File: 240714/합쳐지는 구슬들/merge-marbles.js
function changePosition(n, row, col, direction) {
    const dirs = {'L': [0, -1], 'R': [0, 1], 'U': [-1, 0], 'D': [1, 0]};

    if (direction === 'L' && col === 1) {
        direction = 'R';
    } else if (direction === 'R' && col === n) {
        direction = 'L';
    } else if (direction === 'U' && row === 1) {
        direction = 'D';
    } else if (direction === 'D' && row === n) {
        direction = 'U';
    }

    row += dirs[direction][0];
    col += dirs[direction][1];

    return [row, col, direction];
}

const fs = require("fs");
const input = fs.readFileSync(0).toString().trim().split("\n");
const [n, m, t] = input[0].split(" ").map(Number);

let balls = [];

for (let i = 1; i <= m; i++){
    [r, c, d, w] = input[i].split(" ")
    balls.push({r: Number(r), c: Number(c), d, w: Number(w), num: i})
}

for (let time = 0; time < t; time++){
    const newPositions = new Map();

    for (let ball of balls){
        const [newR, newC, newD] = changePosition(n, ball.r, ball.c, ball.d)
        ball.r = newR
        ball.c = newC
        ball.d = newD

        const key = `${newR},${newC}`
        
        if (newPositions.has(key)){
            newPositions.get(key).push(ball)
        } else {
            newPositions.set(key, [ball])
        }
    }
    
    const newBalls = []
    for (const group of newPositions.values()){
        if (group.length === 1){
            newBalls.push(group[0])
        } else {
            let totalWeight = 0, num = 0, idx = 0
            group.forEach((ball, index) => {
                totalWeight += ball.w
                if (ball.num > num){
                    num = ball.num
                    idx = index
                }
            })
            group[idx].w = totalWeight
            newBalls.push(group[idx])
        }
    }

    balls = newBalls
}

const remainingBallsCount = balls.length
const maxWeight = Math.max(...balls.map((ball) => ball.w))

console.log(remainingBallsCount, maxWeight);

File: 240716/합쳐지는 구슬들/merge-marbles.py
from collections import defaultdict

n, m, t = map(int, input().split())
balls = []
directionDict = {
    'U': 0,
    'R': 1,
    'L': 2,
    'D': 3,
}

for i in range(m):
    r, c, d, w = input().split()
    balls.append({'r':int(r), 'c': int(c), 'd': directionDict[d], 'w': int(w), 'n': i})

def changePosition(r, c, direction):
    dr = [-1, 0, 0, 1]
    dc = [0, 1, -1, 0]

    nr = r + dr[direction]
    nc = c + dc[direction]

    if not (0 < nr <= n and 0 < nc <= n):
        direction = 3 - direction
    else:
        r = nr
        c = nc

    return r, c, direction

for _ in range(t):
    newPositions = defaultdict(list)
    
    for ball in balls:
        newR, newC, newD = changePosition(ball['r'], ball['c'], ball['d'])
        ball['r'] = newR
        ball['c'] = newC
        ball['d'] = newD

        key = f"{newR},{newC}"
        newPositions[key] = newPositions[key] + [ball]

    new_balls = []
    for group in newPositions.values():
        if len(group) == 1:
            new_balls.append(group[0])
        else:
            w = 0
            max_idx = 0
            n = 0
            for idx, ball in enumerate(group):
                w += ball['w']
                if ball['n'] > n:
                    n = ball['n']
                    max_idx = idx
            group[max_idx]['w'] = w
            new_balls.append(group[max_idx])

    balls = [*new_balls]
max_w = 0

for ball in balls:
    max_w = max(max_w, ball['w'])
print(len(balls), max_w)